src/aspire/basis/fle_2d_utils.py

```diff
@@ -0,0 +1,156 @@
+import numpy as np
+import scipy.sparse as sparse
+
+
+def transform_complex_to_real(B_conj, ells):
+    """
+    Transforms coefficients of the matrix B (see Eq. 3) from complex
+        to real. B is the linear transformation that takes FLE coefficients
+        to images.
+
+    :param B_conj: Complex conjugate of the matrix B.
+    :param ells: List of ells (Bessel function orders) in this basis.
+    :return: Transformed matrix.
+    """
+    num_basis_functions = B_conj.shape[1]
+    X = np.zeros(B_conj.shape, dtype=np.float64)
+
+    for i in range(num_basis_functions):
+        ell = ells[i]
+        if ell == 0:
+            X[:, i] = np.real(B_conj[:, i])
+        # for each ell != 0, we can populate two entries of the matrix
+        # by taking the complex conjugate of the ell with the opposite sign
+        if ell < 0:
+            s = (-1) ** np.abs(ell)
+            x0 = (B_conj[:, i] + s * B_conj[:, i + 1]) / np.sqrt(2)
+            x1 = (-B_conj[:, i] + s * B_conj[:, i + 1]) / (1j * np.sqrt(2))
+            X[:, i] = np.real(x0)
+            X[:, i + 1] = np.real(x1)
+
+    return X
+
+
+def precomp_transform_complex_to_real(ells):
+
+    count = len(ells)
+    num_nonzero = np.sum(ells == 0) + 2 * np.sum(ells != 0)
+    idx = np.zeros(num_nonzero, dtype=int)
+    jdx = np.zeros(num_nonzero, dtype=int)
+    vals = np.zeros(num_nonzero, dtype=np.complex128)
+
+    k = 0
+    for i in range(count):
+        ell = ells[i]
+        if ell == 0:
+            vals[k] = 1
+            idx[k] = i
+            jdx[k] = i
+            k = k + 1
+        if ell < 0:
+            s = (-1) ** np.abs(ell)
+
+            vals[k] = 1 / np.sqrt(2)
+            idx[k] = i
+            jdx[k] = i
+            k = k + 1
+
+            vals[k] = s / np.sqrt(2)
+            idx[k] = i
+            jdx[k] = i + 1
+            k = k + 1
+
+            vals[k] = -1 / (1j * np.sqrt(2))
+            idx[k] = i + 1
+            jdx[k] = i
+            k = k + 1
+
+            vals[k] = s / (1j * np.sqrt(2))
+            idx[k] = i + 1
+            jdx[k] = i + 1
+            k = k + 1
+
+    A = sparse.csr_matrix((vals, (idx, jdx)), shape=(count, count), dtype=np.complex128)
+
+    return A
+
+
+def barycentric_interp_sparse(x, xs, ys, s):
+    """
+    Perform barycentric interpolation to compute values of Betas at the points
+        `x`, based on their values (`ys`) at known points (`xs`).
```

Comment:
Returns the sparse matrices that perform barycentric interpolation to compute values of Betas at the points `x`, based on their values (`ys`) at known points (`xs`), and the transpose of this operation. For each target point in x, only s centered source points from xs around the target point are used.